### ISSUE TITLE
Executing "Content" step before "ContentDefinition" caused null reference exceptions

### ISSUE THREAD

**Issue by MikeAlhayek** [OrchardCMS/OrchardCore#12566]
### Describe the bug
The recipe steps are executed in the order they are listed in the the recipe file. so if "Content" step is listed above "ContentDefinition" when it's executed, there could be no content definition which causes a null reference exception. 

### To Reproduce
1. Import the following recipe in a site that had no content type definition named `HomePage` 
```
{
  "steps": [

     // Content
    {
      "name": "Content",
      "data": [
        // Default home page
        {
          "ContentItemId": "4dcz531ajab690tn643jar5rw7",
          "ContentType": "HomePage",
          "DisplayText": null,
          "Latest": true,
          "Published": true,
          "Owner": "[js: parameters('AdminUserId')]",
          "Author": "[js: parameters('AdminUsername')]",
          "HomePage": {},
          "AutoroutePart": {
            "Path": "home-page",
            "SetHomepage": false,
            "Disabled": false,
            "RouteContainedItems": false,
            "Absolute": false
          },
          "HtmlBodyPart": {
            "Html": "<p>Test Page</p>"
          }
        }
      ]
    },


    // Content Definitions
    {
      "name": "ContentDefinition",
      "ContentTypes": [
        {
          "Name": "HomePage",
          "DisplayName": "Landing Page",
          "Settings": {
            "ContentTypeSettings": {
              "Listable": true,
              "Draftable": true,
              "Versionable": true,
              "Securable": true
            },
            "FullTextAspectSettings": {}
          },
          "ContentTypePartDefinitionRecords": [
            {
              "PartName": "HomePage",
              "Name": "HomePage",
              "Settings": {
                "ContentTypePartSettings": {
                  "Position": "0"
                }
              }
            },
            {
              "PartName": "AutoroutePart",
              "Name": "AutoroutePart",
              "Settings": {
                "ContentTypePartSettings": {
                  "Position": "1"
                },
                "ContentIndexSettings": {},
                "AutoroutePartSettings": {
                  "AllowCustomPath": true
                }
              }
            },
            {
              "PartName": "HtmlBodyPart",
              "Name": "HtmlBodyPart",
              "Settings": {
                "ContentTypePartSettings": {
                  "Position": "2"
                },
                "ContentIndexSettings": {},
                "HtmlBodyPartSettings": {
                  "SanitizeHtml": false
                }
              }
            }
          ]
        }
      ],
      "ContentParts": []
    }
  ]
}
```

I think recipe steps should be executed in a logical order to ensure dependencies step are executed first.

For example, we should execute the following key steps in the following order
1. Feature Step
2. Content Definition
3. Content
4. ReplaceContent
5. DeleteContent

**Comment by sebastienros** [comment]
I suggest to create a new service/interface that is executed before the IRecipeStepHandlers, it would get the recipe as-is and be able to change it, in this case by moving all ContentSteps after ContentDefinitionSteps. It could also be able to remove passwords, change position based on data, create new steps dynamically, ...

These handlers might have to be executed in a loop until no changes are made. For instance if one handler changes the order or two steps, but a previous handler would have changed it too. And if there is an infinite loop (limited number of retries) then we have found an issue in a recipe that can't be resolved.

**Comment by jtkech** [comment]
We already discussed about this #4393, so as I remember.

- For this special case of `ContentStep` before `ContentDefinitionStep`, we already fixed it by executing the related items import in a Deferred Task. For now it is only applied if the shell is not Activated e.g. during a setup, but would be easy to also apply it on an already activated tenant.

- Still for the same case, not fully sure but changing the order of the steps may not be sufficient if some needed Tables are not yet created, that's why we used a deferred task.

- @MikeAlhayek adding an `Order` prop to handlers in #12568 is not sufficient, each step call all handlers, yes only the ones related to a given step are doing something, but steps are still executed in the same order.

    Then we could still think about using handlers to deduce the steps order, but each handler would need to provide its related step name, also multiple handlers may be related to the same step. So maybe the registration order of these handlers, according to features dependencies, may help more, handlers collaborating to the same step would need to depend on the module/feature that defines the step.

- At some point I thought about using as possible existing dependency ordering, would not work for this case as `ContentTypes` depends on `Contents`, `ContentTypes` is an UI but maybe registers services needed by `Contents`, if so would need another "core" module on which both would depend.

    But anyway as said above, for this specific case we needed to use a deferred task.

- Finally what described @sebastienros is like a graph to be resolved, as we do with the features dependency graph to resolve "recursively" the features ordering.

**Comment by MikeAlhayek** [comment]
@jtkech what is the solution then? We should be able to deferred task that can still run after ContentStep. For example, I want to execute some code that depends on default content. So the code would need to real the `ContentItemIndex` table or `Users` table and made some sort of changes.

I don't think adding handler as @sebastienros suggested here would fix the issue of needed to defer execution until after all migrations are executed first.

**Comment by jtkech** [comment]
@MikeAlhayek Yes many things so some unordered thinkings as I remember ATM.

Yes there is the steps ordering, but also for a given step the order of recipe step handlers, here handlers are already resolved in the order they are registered so based on the features dependencies. So if you have a custom recipe defined in a feature that depends on `OC.Contents` it will be executed after the default.

Yes creating contents may still need that some definitions are already persisted, but it may depend knowing that as soon we read an item from YesSql there is an auto flush meaning the cmds are executed, here the transaction is not yet committed but it is configured to allow to read uncommitted data.

That said, if deferred tasks are still needed, they are executed in the order they are enlisted, so defining a step handler acting on the ContentStep in a feature that depends on Contents will also ensure that its deferred tasks will be executed after the default one enlisted by the default content step handler.

---

Then regarding the order of the steps themselves, maybe here we could also use the inner dependencies ordering, anyway we saw that this is `OC.ContentTypes` that depends on `OC.Contents`, about this see #12617 and #12618 where I suggest to use the opposite dependency.

Hmm, if recipe step handlers also provide a `Name`, if we apply #12618 and dependencies are respected, when resolving all step handlers normally those related to ContentStep would be all after those related to type definitions. So by applying a distinct by step Name we could get an ordered list of step names.

Then maybe an Order property to override the default for specific cases, as we do for startups that are executed in an order based on feature dependencies unless for those overriding their Order property.

---

Hmm finally I think there are 2 separated things, the order that step handlers can collaborate for a given step, which is already based on feature dependencies which is good. Then a separate way to define the order of steps, for now from the recipes themselves, but not through handlers, maybe new registered `IRecipeStepName` exposing a Name and maybe an Order to override the default ordering.

**Comment by sebastienros** [comment]
> maybe here we could also use the inner dependencies ordering

Won't work, not related.

My suggestion will work. And the implementation of this new service need to use a stable sort, meaning it should not change the order of steps if it doesn't care about them. Would also ensure idempotency (running the same implementation twice won't change the order) so we can _eventually_ get to a stable order once all the implementations have been executed (at least once).

And this is not incompatible with deferred tasks. Worst case these would also need to be ordered, with the same service implementations.

NB: these implementations will usually know how to order two specific steps only.

**Comment by sebastienros** [comment]
If the issue is not about ordering steps, please repeat it simply.

**Comment by sebastienros** [comment]
If recurring recipes is an issue we might need to flatten the steps before ordering.

**Comment by jtkech** [comment]
> Won't work, not related.

Yes, I thought that it could be used as a first ordering but finally as said in the same comment.

> > Hmm finally I think there are 2 separated things, the order that step handlers can collaborate for a given step, which is already based on feature dependencies which is good. Then a separate way to define the order of steps, for now from the recipes themselves, bla bla ...

Just for info, for now deferred tasks are executed in the order they are enlisted, so normally in the same order that steps triggering them are finally ordered.

> these implementations will usually know how to order two specific steps only.

Interesting, each one defining a simple dependency to build at the end a flat dependency "graph".

IMHO The way steps are currently ordered by recipes themselves is sufficient, at least for now.

**Comment by sebastienros** [comment]
@jtkech do you think the recipe provided in the issue is valid? Do you think my suggestion would fix it? At least not taking sub-recipes into account as they would be also re-ordered (in their own scope).

**Comment by jtkech** [comment]
Okay no problem.

Maybe I was thinking about too many cases and too lazy to take them into account ;) This looking at the comments of for example #9046, #8785, #8967, and thinking about the related history.

Yes recipes can be executed during a setup or not, we introduced migrations recipes to run for example `ContentDefinitionStep` but (as other recipes) that may also contain a `ContentStep`, in that case the step is still run in its scope but the parent scope may have other needed regular migrations that are not yet completed, that's why we used a deferred task for `ContentStep`.

So only re-ordering steps fix this issue but would have not fixed the above case. Note: `ContentStep` only use a deferred task during a setup but easy to always use it which would also fix this issue.

And here `ContentDefinitionStep` can be run (as any step) in a non migration recipe, maybe we allow too many things. That said most of the time this kind of issue is related to `ContentDefinitionStep` and/or `ContentStep` and/or `FeatureStep`, because a tenant should have the right `ShellDescriptor` and the right database schema before executing anything else.

Yes your suggestion looks good and may cover more use cases, but just ideas on the fly by first more focusing on what the system does to activate a tenant (and in which order) before it can be used, so maybe we just need to first run `FeatureSteps`, then `ContentDefinitionSteps`, there are different ways to do it but maybe not worth it if it's not less complex than your suggestion ;)

Finally we could also say, as when we are writing code, that in a recipe we should define features and types definitions before using them in a following step.
